### PR TITLE
Fix #19823. Disallow overriding with wrong object type

### LINKED ISSUE #19823: Custom scenery overrides cause glitching when the objects are different scenery types
### Operating System

Windows 10, 64-bit

### OpenRCT2 build

OpenRCT2, v0.4.4-38-g50242c2 (50242c2 on develop) provided by GitHub

### Base game

RollerCoaster Tycoon 2

### Area(s) with this issue?

This issue has to do with park objects and/or asset packs

### Describe the issue

If a custom object in parkobj format is small scenery, but overrides a DAT object that is large scenery, it causes graphical glitches where other unrelated objects get replaced, seemingly at random.

It makes sense that replacing objects of different types wouldn't work, but IMO the game should just ignore these objects, if that's possible.

### Steps to reproduce

1. Clear your object folder, and download the DKSO_Workbench_2023e bench (see attachments below)
2. Download The Lonely Planet from NE [here](https://www.nedesigns.com/park/5680/the-lonely-planet/).
3. Open the workbench to get its objects installed to your object folder.
4. Open The Lonely Planet.
5. Note that some random bushes have been replaced with trees.

This is because there is a custom tree object (fidwell.scenery_small.tl2) that is small scenery, which overwrites GWTRL2, which is large scenery. Neither object is supposed to be in the affected area, yet some glitchy trees show up anyway.

Expected:

Actual:

### Attachments

[DKSO_Workbench_2023e.zip](https://github.com/OpenRCT2/OpenRCT2/files/11150635/DKSO_Workbench_2023e.zip)

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Some community objects have been made incorrectly and override small scenery objects with large scenery objects. This prevents that overload from being allowed.

### CODE REVIEW

**Comment by duncanspumpkin** [comment]
Just thinking i might not have to be as harsh as to force the object to fail. Could just blank out the id.

**Comment by ZehMatt** [comment]
> Just thinking i might not have to be as harsh as to force the object to fail. Could just blank out the id.

I don't think its bad, errors should be obvious, it could be a in-game window but we don't have that atm.

**Comment by ZehMatt** [comment]
> Just thinking i might not have to be as harsh as to force the object to fail. Could just blank out the id.

I don't think its bad, errors should be obvious, it could be a in-game window but we don't have that atm.

**Comment by fidwell** [comment]
I'd like a chance to test this out tonight, before it gets merged. I have a lot of test cases to verify.

**Comment by duncanspumpkin** [comment]
> Can you provide such sample object?

There are some sample objects in the linked issue

**Comment by fidwell** [comment]
I can provide some specific single objects here in a bit

**Comment by fidwell** [comment]
Here is a sample scenario. GWTRJT1.DAT is a recolorable jungle tree, which was created as a large object. The parkobj is a small object that attempts to override the DAT. If you make a save file with the DAT tree in it, save it, then add the parkobj to your object folder and re-load the save, the tree won't appear, and the game will throw an error when you quit.
[TJT1.zip](https://github.com/OpenRCT2/OpenRCT2/files/11456339/TJT1.zip)

**Comment by duncanspumpkin** [comment]
> > Just thinking i might not have to be as harsh as to force the object to fail. Could just blank out the id.
> 
> I don't think its bad, errors should be obvious, it could be a in-game window but we don't have that atm.

I tried out making it just zero out the descriptor and that also works. But if i add an error message you get quite a bit of spam from it. 

On further inspection this is the source of tons of crashes suggest we keep this as object is invalid.

**Comment by duncanspumpkin** [comment]
Likely fixes: #20178, #20163, #20151, #20144, #20132, #20101, #20098, #20090, #20079, #20068, #19966, #19912, #19884, #19804, #19786, #19677, #19658

**Comment by fidwell** [comment]
There is a bizarre side effect with scenery _group_ objects. When I try to load a park that has a scenery group parkobj that overrides, I get that new error message, and a new parkobj gets created in my object folder with `-02` tacked on to the end of the filename.

**Comment by duncanspumpkin** [comment]
> There is a bizarre side effect with scenery _group_ objects. When I try to load a park that has a scenery group parkobj that overrides, I get that new error message, and a new parkobj gets created in my object folder with `-02` tacked on to the end of the filename. If I try again, then it creates `-03.parkobj`, etc.
> 
> Going into the debugger, the ObjectFactory thinks the object type of the original tab group is Ride (0) instead of Scenery Group (7). I guess that is something we have to work around?

Hmm think i'll have to switch to the zeroing out the descriptor approach. That prevents the duplicate creation from the scenery groups.

**Comment by duncanspumpkin** [comment]
Amusingly after doing this we now have a save that packs unrelated objects into it. The fidwell objects that have the bad override are not classed as part of the save but they are packed with the save.

**Comment by fidwell** [comment]
This does pass my first two test cases, though. The basic jungle tree scenario I mentioned above, and the Lonely Planet save from the original issue, both now work as intended.

**Comment by duncanspumpkin** [comment]
> This does pass my first two test cases, though. The basic jungle tree scenario I mentioned above, and the Lonely Planet save from the original issue, both now work as intended.

Please retest with the new version. I've switched it to just removing the override rather than black listing the object.

**Comment by fidwell** [comment]
Yeah this is a better approach, at least from my perspective. It is unfortunate that my scenery group overrides won't work anymore, but it's a sacrifice worth making I guess.

**Review by fidwell — APPROVED** [review]
Needs a changelog entry but otherwise LGTM

**Review by ZehMatt — APPROVED** [review]

**Comment by ZehMatt** [comment]
Also probably worth bumping the network version.

File: src/openrct2/object/ObjectFactory.cpp
Comment (duncanspumpkin) [reply src/openrct2/object/ObjectFactory.cpp:546]: Yes it is a nice little bonus that means that this bug doesn't cause as much damage as I initially thought.

**Review by Gymnasiast — APPROVED** [review]

File: src/openrct2/object/ObjectFactory.cpp
Comment (fidwell) [reply src/openrct2/object/ObjectFactory.cpp:546]: Yeah, I wonder if this has been the cause of behavior we've seen, where the game requires a user to have an overridden object in their objects folder, even though a parkobj in the park they're loading should supercede it. We'll see!

**Pull request merged**
